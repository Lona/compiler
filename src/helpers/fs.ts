import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
