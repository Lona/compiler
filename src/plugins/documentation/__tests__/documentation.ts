import fs from 'fs'
import path from 'path'
import { createFs, copy } from 'buffs'
import plugin from '../index'
import Helpers from '../../../helpers'

const workspacePath = path.join(__dirname, '../../../../examples/workspace')

it('converts workspace', async () => {
  const source = createFs()

  copy(fs, source, workspacePath, '/')

  const helpers = Helpers(source, '/')

  await plugin.convertWorkspace('/', helpers, { output: '/docs.json' })

  const output = JSON.parse(source.readFileSync('/docs.json', 'utf8'))

  expect(output).toMatchSnapshot()
})
