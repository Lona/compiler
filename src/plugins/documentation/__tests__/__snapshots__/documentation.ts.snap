// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts workspace 1`] = `
Object {
  "files": Array [
    Object {
      "contents": Object {
        "type": "documentationPage",
        "value": Object {
          "children": Array [],
          "mdxString": "# Colors

<div class=\\"lona-token lona-token-color\\">
  <div  class=\\"lona-token-preview lona-token-preview-color\\" data-color=\\"#45CBFF\\"></div>
  <div class=\\"lona-token-details lona-token-details-color\\">
    <span class=\\"lona-token-name lona-token-name-color\\">primary</span>
    <span class=\\"lona-token-value lona-token-value-color\\">#45CBFF</span>
  </div>
</div>

<div class=\\"lona-token lona-token-color\\">
  <div  class=\\"lona-token-preview lona-token-preview-color\\" data-color=\\"#45CBFF\\"></div>
  <div class=\\"lona-token-details lona-token-details-color\\">
    <span class=\\"lona-token-name lona-token-name-color\\">accent</span>
    <span class=\\"lona-token-value lona-token-value-color\\">#45CBFF</span>
  </div>
</div>

<div class=\\"lona-token lona-token-color\\">
  <div  class=\\"lona-token-preview lona-token-preview-color\\" data-color=\\"#29D7FF\\"></div>
  <div class=\\"lona-token-details lona-token-details-color\\">
    <span class=\\"lona-token-name lona-token-name-color\\">testSaturate</span>
    <span class=\\"lona-token-value lona-token-value-color\\">#29D7FF</span>
  </div>
</div>",
        },
      },
      "inputPath": "Colors.md",
      "name": "Colors",
      "outputPath": "Colors.mdx",
    },
    Object {
      "contents": Object {
        "type": "documentationPage",
        "value": Object {
          "children": Array [
            "Colors.md",
            "TextStyles.md",
            "Shadows.md",
          ],
          "mdxString": "# Flat Tokens

This is an example workspace for testing token generation.

### Pages

<a class=\\"page\\" href=\\"Colors.md\\">Colors</a>

<a class=\\"page\\" href=\\"TextStyles.md\\">TextStyles</a>

<a class=\\"page\\" href=\\"Shadows.md\\">Shadows</a>",
        },
      },
      "inputPath": "README.md",
      "name": "README",
      "outputPath": "README.mdx",
    },
    Object {
      "contents": Object {
        "type": "documentationPage",
        "value": Object {
          "children": Array [],
          "mdxString": "<div class=\\"lona-token lona-token-shadow\\">
  <div  class=\\"lona-token-preview lona-token-preview-shadow\\" data-x=\\"0\\" data-y=\\"2\\" data-blur=\\"2\\" data-radius=\\"0\\" data-color=\\"#45CBFF\\"></div>
  <div class=\\"lona-token-details lona-token-details-shadow\\">
    <span class=\\"lona-token-name lona-token-name-shadow\\">small</span>
    <span class=\\"lona-token-value lona-token-value-shadow\\">0px 2px 2px 0px #45CBFF</span>
  </div>
</div>",
        },
      },
      "inputPath": "Shadows.md",
      "name": "Shadows",
      "outputPath": "Shadows.mdx",
    },
    Object {
      "contents": Object {
        "type": "documentationPage",
        "value": Object {
          "children": Array [],
          "mdxString": "<div class=\\"lona-token lona-token-textStyle\\">
  <div  class=\\"lona-token-preview lona-token-preview-textStyle\\" data-fontFamily=\\"Helvetica\\" data-fontWeight=\\"700\\" data-fontSize=\\"28\\" data-color=\\"teal\\"></div>
  <div class=\\"lona-token-details lona-token-details-textStyle\\">
    <span class=\\"lona-token-name lona-token-name-textStyle\\">heading1</span>
    <span class=\\"lona-token-value lona-token-value-textStyle\\">Helvetica 700 28px teal</span>
  </div>
</div>",
        },
      },
      "inputPath": "TextStyles.md",
      "name": "TextStyles",
      "outputPath": "TextStyles.mdx",
    },
  ],
  "flatTokensSchemaVersion": "0.0.1",
}
`;
